lib/Dialect/TritonGPU/Transforms/Pipeliner/PipelineExpander.cpp

```diff
@@ -291,7 +295,19 @@ LogicalResult LoopPipelinerInternal::emitPrologue(RewriterBase &rewriter) {
        llvm::zip(forOp.getRegionIterArgs(), forOp.getInitsMutable())) {
     setValueMapping(arg, operand.get(), 0);
   }
+
+  // If the incoming value to an iter arg from the loop yield is defined outside
+  // the loop, then that means the iter arg takes that value for all stages
+  // after the first stage.
   auto yield = cast<scf::YieldOp>(forOp.getBody()->getTerminator());
+  for (auto [arg, operand] :
+       llvm::zip(forOp.getRegionIterArgs(), yield->getOpOperands())) {
+    if (forOp.getBodyRegion().isAncestor(operand.get().getParentRegion()))
+      continue;
+    for (int64_t i = 1; i < maxStage; ++i)
+      setValueMapping(arg, operand.get(), i);
+  }
```

Comment:
interesting, does that happen in real cases?

Reply:
Yes, something like

```python
a = 42
for i in range(...):
    use(a)
    a = 27
```

Would trigger this.

I didn't hit this literally in real code, but some combination of transformations caused this to happen.